circle/src/deep_quotient.rs

```diff
@@ -0,0 +1,101 @@
+use alloc::vec::Vec;
+
+use itertools::{izip, Itertools};
+use p3_commit::PolynomialSpace;
+use p3_field::extension::{Complex, ComplexExtendable, ExtensionPowersReducer};
+use p3_field::{batch_multiplicative_inverse, ExtensionField};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::Matrix;
+use p3_maybe_rayon::prelude::*;
+use p3_util::log2_strict_usize;
+use tracing::instrument;
+
+use crate::domain::CircleDomain;
+use crate::util::v_n;
+
+/// Compute numerator and denominator of the "vanishing part" of the DEEP quotient
+/// Section 6, Remark 21 of Circle Starks (page 30 of first edition PDF)
+/// Re(1/v_gamma) + alpha^L Im(1/v_gamma)
+/// (Other "part" is \bar g - \bar v_gamma)
+pub(crate) fn deep_quotient_vanishing_part<F: ComplexExtendable, EF: ExtensionField<F>>(
+    x: Complex<F>,
+    zeta: Complex<EF>,
+    alpha_pow_width: EF,
+) -> (EF, EF) {
+    let x_rotate_zeta: Complex<EF> = x.rotate(zeta.conjugate());
+
+    let v_gamma_re: EF = EF::one() - x_rotate_zeta.real();
+    let v_gamma_im: EF = x_rotate_zeta.imag();
+
+    (
+        v_gamma_re - alpha_pow_width * v_gamma_im,
+        v_gamma_re.square() + v_gamma_im.square(),
+    )
+}
+
+pub(crate) fn deep_quotient_reduce_row<F: ComplexExtendable, EF: ExtensionField<F>>(
+    alpha_reducer: &ExtensionPowersReducer<F, EF>,
+    x: Complex<F>,
+    zeta: Complex<EF>,
+    ps_at_x: &[F],
+    ps_at_zeta: &[EF],
+) -> EF {
+    let (vp_num, vp_denom) =
+        deep_quotient_vanishing_part(x, zeta, alpha_reducer.factor.exp_u64(ps_at_x.len() as u64));
+    vp_num
+        * vp_denom.inverse()
+        * (alpha_reducer.reduce_base(ps_at_x) - alpha_reducer.reduce_ext(ps_at_zeta))
```

Comment:
alpha_reducer is a dot product right? In that case, I think it should be cheaper to subtract before the dot product. 

E.g. something of the form (I know this doesn't work as written but hopefully the meaning is clear):

alpha_reducer.reduce_base(ps_at_x - ps_at_zeta)

This replaces N <EF, F> muls by N <EF, F> subtracts which should be much quicker.

Reply:
Yep, nice catch. Was just copying this from `deep_quotient_reduce_matrix`, it only gets called a few times